docs/User/Tutorials/haskell-lazy-evaluation.md

```diff
@@ -0,0 +1,171 @@
+# Haskell Lazy Evaluation
+
+Haskell is fundamentally different from other languages. In most
+languages, data has a concrete representation in memory.  In Haskell,
+data may have a concrete representation in memory or it could be a
+reference to a computation or expression which when evaluated would
+produce concrete data.
+
+The entire program in Haskell is an unevaluated expression which is
+evaluated on-demand as needed. The expression is forced to produce data
+when result has to be printed on console or stored on disk or sent over
+network, or the programmer wants to force convert it. The conversion of
+expressions to concrete data can be controlled by the programmer.
+
+## Unevaluated Expressions
+
+When we assign a value to a name, the name refers to the unevaluated
+expression and not the result of evaluating that expression. For example:
+
+```
+>>> infiniteList = [1..]
+```
+
+The name `infiniteList` refers to an expression which when evaluated
+would generate the list. `infiniteList` is not an actual physical list
+but just a formula to generate the list `when needed`. That is the
+reason we can represent an infinite list without requiring infinite
+amount of memory.
+
+Let's try to get the length of this list:
+
+```
+>>> len = length infiniteList
+```
+
+The name `len` refers to an expression that when evaluated would give us
+the length of the list.  Till now Haskell has been saying that I duly
+noted how to do what you want to do but I will do it only when you need
+it. Let's ask it to print `len` now:
+
+```
+>>> len
+...hangs forever
+```
+
+Hard pressed, Haskell now has no choice but to start evaluating the
+expression but we know it can never finish because the list is infinite.
+
+All expressions in Haskell are unevaluated by default.  Haskell
+evaluates the expressions only when it is absolutely required e.g. when
+we have to print something, or write it to a file or to network.
+
+## On-demand Evaluation
```

Comment:
The title does not seem to justify the content. We are talking about memory.

Reply:
I changed the content a little bit.